### ISSUE TITLE
Hermes Chain Configuration Unclear

### ISSUE THREAD

**Issue by alexanderbez** [informalsystems/ibc-rs#1272]
## Crate

ibc-relayer(-cli)

## Summary

Please close this issue if a duplicate.

Relayers are still a relatively new thing in the Cosmos/IBC landscape and there aren't many operators that run them or know how to run and configure them effectively.

The current [configuration](https://github.com/informalsystems/ibc-rs/blob/v0.6.2/config.toml), specifically the `[[chain]]`, doesn't really document what chains the values correspond to.

Specifically, I imagine a relayer will listen for events/txs from a _source_ chain and relay them to a _destination_ chain and this relayer will typically exist on the same host of the _destination_ chain. It's not clear, for example, if the `id` is the _source_ chain in this context or the chain being relayed (txs sent) to (I assume the _destination_ chain). The same applies for the gRPC, websocket, and RPC endpoints. I think the rest of the configurations are pretty self-explanatory (they're for the _destination_ chain).

## Proposal

Document in the [configuration](https://github.com/informalsystems/ibc-rs/blob/v0.6.2/config.toml) if the relevant values correspond to the _source_ or _destination_ chain.

____

#### For Admin Use

- [ ] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate milestone (priority) applied
- [ ] Appropriate contributors tagged
- [ ] Contributor assigned/self-assigned

**Comment by alexanderbez** [comment]
Based on the example config, https://hermes.informal.systems/config.html#example-configuration-file, it seems Hermes expects an operator to run a full-node for _both_  chains? If so, feel free to close this issue :)

**Comment by adizere** [comment]
Hi Alexandr,
Thanks for opening this issue, I realize our documentation is not very clear in the points you raised! I'll try to clarify here, and please follow-up if I'm missing anything.

> Based on the example config, https://hermes.informal.systems/config.html#example-configuration-file, it seems Hermes expects an operator to run a full-node for _both_ chains, typically on the same host? If so, feel free to close this issue :)

Hermes does expect connectivity to all endpoints (namely, rpc, grpc, and ws) for all chains specified in config.toml. But it depends what functionality do you want to use Hermes for concretely. Suppose you configure two chains `cosmoshub-4` and `osmosis-1` in config.toml, but all you ever use Hermes is to


```
hermes query clients cosmoshub-4
```

Then at runtime Hermes will not try to establish connection to any of the endpoints configured for chain `osmosis-1`.

>typically on the same host?

Not necessarily on the same host.


>The current configuration, specifically the [[chain]], doesn't really document what chains the values correspond to.

Each `[[chain]]` section has a an [id tag](https://github.com/informalsystems/ibc-rs/blob/6be938cfd5b5e5d394172989f8078d5893bec7d5/config.toml#L51), which specifies the chain identifier for that whole section.

> Specifically, I imagine a relayer will listen for events/txs from a source chain and relay them to a destination chain and this relayer will typically exist on the same host of the destination chain.

If you run `hermes start`, then Hermes will relay on _all_ channels available between _all_ the configured chains. In this way, every configured chain will act as a source (in the sense that Hermes listens for events) and as a destination (to relay packets that others chains have sent). To restrict relaying on specific channels, you can use [packet_filter policies](https://github.com/informalsystems/ibc-rs/blob/6be938cfd5b5e5d394172989f8078d5893bec7d5/config.toml#L130-L133).

**Comment by alexanderbez** [comment]
Thanks for the excellent explanation @adizere!

Quick clarification,

>  In this way, every configured chain will act as a source (in the sense that Hermes listens for events) and as a destination (to relay packets that others chains have sent). To restrict relaying on specific channels, or uni-directionally, you can use packet_filter policies.

- Every `chain` will act as source of events, but what does it do with those events? I imagine it constructs messages to sign and relay to other chains, right? So it'll need configuration for the destination chain(s) as well?
- Every `chain` will acts as a destination, but it only for txs that the relayer signs and constructs from another _source_ `chain`, right?

**Comment by adizere** [comment]
> * Every `chain` will act as source of events, but what does it do with those events? I imagine it constructs messages to sign and relay to other chains, right? So it'll need configuration for the destination chain(s) as well?
> * Every `chain` will acts as a destination, but it only for txs that the relayer signs and constructs from another _source_ `chain`, right?

That's correct. If there are only two chains configured, then only packets between those two can Hermes relay, i.e., the two chains will serve as a source for each other, and likewise as a destination for each other's relevant events. Hermes will ignore all events that pertain to chains which are unknown (unspecified in config.toml).

**Comment by adizere** [comment]
Glad to help!

Is there a specific use-case you might have in mind that you'd like Hermes to cover (e.g., uni-directional relaying)? Also, I leave it up to you to close this issue if we managed to clarify it.

**Comment by alexanderbez** [comment]
This clarifies it for me. Thank you so much.

**Comment by romac** [comment]
Let's actually re-open this until we clarify the points above directly in the configuration/guide.
